src/js/background.js

```diff
@@ -0,0 +1,214 @@
+      for (let j = games.length - 1; j >= 0; j--) {
+        if (games[j].url === gameUrl) {
+          return games[j].pgn;
+        }
+      }
+    }
+  } catch (error) {
+    sendLogMessage(error, tabId);
+    return null;
+  }
+
+  return null;
+};
+
+const getPGNManual = async (tabId) => {
+  await waitAndClick(".share");
+  await waitAndClick(".board-tab-item-underlined-component");
+  await waitAndClick(".share-menu-tab-pgn-toggle");
+
+  if (await waitForElement("[name='pgn']")) {
+    const [pgn] = await browser.tabs.executeScript(tabId, {
+      code: `document.querySelector("[name='pgn']").value;`,
+    });
+    await waitAndClick("[data-cy='share-menu-close']");
+
+    return pgn;
+  }
+
+  return null;
+};
+
+const clickElement = async (querySelector, tabId) => {
+  await browser.tabs.executeScript(tabId, {
+    code: `document.querySelector("${querySelector}").click();`,
+  });
+};
+
+const waitForElement = async (
+  querySelector,
+  tabId,
+  timeLeft = 5000,
+  retryDelay = 100
+) => {
+  const elementFoundScript = `document.querySelector("${querySelector}") !== null;`;
+
+  let isElement = null;
+
+  while (!isElement && timeLeft > 0) {
+    await new Promise((r) => setTimeout(r, retryDelay));
+    timeLeft -= retryDelay;
+    [isElement] = await browser.tabs.executeScript(tabId, {
+      code: elementFoundScript,
+    });
+  }
+
+  if (!isElement)
+    sendLogMessage(`Cannot find \`${querySelector}\` element!`, tabId);
+
+  return isElement;
+};
+
+const waitAndClick = async (querySelector, tabId) => {
+  if (await waitForElement(querySelector, tabId))
+    await clickElement(querySelector, tabId);
+};
+
+const lichessAnalyse = async (tabId, pgn) => {
+  if (await waitForElement("[name='analyse']", tabId)) {
+    const [loggedIn] = await browser.tabs.executeScript(tabId, {
+      code: `!document.querySelector("[name='analyse']").disabled;`,
+    });
+    if (loggedIn) await waitAndClick("[name='analyse']", tabId);
+
+    if (await waitForElement("[name='pgn']", tabId)) {
+      await browser.tabs.executeScript(tabId, {
+        code: `document.querySelector("[name='pgn']").value = \`${pgn}\`;`,
+      });
+      await waitAndClick(".submit", tabId);
+
+      /*
+       * Bugfix where status was loading and firefox throwed
+       * an error about invalid host permissions
+       */
+      let status = "loading";
+      while (status === "loading") {
+        let getting = await browser.tabs.get(tabId);
+        status = getting.status;
+      }
+
+      if (await waitForElement("#analyse-toggle-ceval", tabId)) {
+        const [localEval] = await browser.tabs.executeScript(tabId, {
+          code: `document.querySelector("#analyse-toggle-ceval").checked;`,
+        });
+        if (!localEval)
+          await waitAndClick("[for='analyse-toggle-ceval']", tabId);
+      }
+    }
+  }
+};
+
+/*
+ * Structure:
+ * {
+ *   tabId0: bool,
+ *   tabId1: bool
+ * }
+ */
```

Comment:
This is something you could encode as a JSDoc comment. This would help tsserver understand the code (if you're using VSCode, you will get more intellisense if you use JSDoc even in a JS file).